cas-mfa-java/src/main/java/net/unicon/cas/mfa/authentication/RegisteredServiceMfaRoleProcessor.java

```diff
@@ -0,0 +1,254 @@
+                        serviceMfaData.getAttributeName(),
+                        serviceMfaData.getAuthenticationMethod());
+                return null;
+            }
+            final int mfaMethodRank = this.authenticationMethodConfiguration.getAuthenticationMethod(
+                    serviceMfaData.getAuthenticationMethod()).getRank();
+            final MultiFactorAuthenticationSupportingWebApplicationService svc =
+                    this.mfaServiceFactory.create(targetService.getId(), targetService.getId(),
+                            targetService.getArtifactId(), serviceMfaData.getAuthenticationMethod(),
+                            MultiFactorAuthenticationSupportingWebApplicationService.AuthenticationMethodSource.PRINCIPAL_ATTRIBUTE);
+
+            return new MultiFactorAuthenticationRequestContext(svc, mfaMethodRank);
+        }
+
+        logger.trace("{} did not match {}", attributeValue, serviceMfaData.getAttributePattern());
+        return null;
+    }
+
+    /**
+     * Match will compare the value to the pattern.
+     * @param attributePattern the pattern to check
+     * @param attributeValue the value to check
+     * @return true if a match is found. otherwise false
+     */
+    private boolean match(final String attributePattern, final String attributeValue) {
+        Pattern pattern = translationMap.get(attributePattern);
+        if (pattern == null) {
+            pattern = Pattern.compile(attributePattern);
+        }
+
+        return pattern.matcher(attributeValue).matches();
+    }
+
+    /**
+     * Looks up the mfa data for a specific service.
+     * @param targetService the service to check
+     * @return service specific mfa settings
+     */
+    private ServiceMfaData getServicesAuthenticationData(final WebApplicationService targetService) {
+        final RegisteredService registeredService = this.servicesManager.findServiceBy(targetService);
+        if (registeredService == null) {
+            logger.debug("No registered service is found. Delegating to the next argument extractor in the chain...");
+            return null;
+        }
+
+        if (!(registeredService instanceof RegisteredServiceWithAttributes)) {
+            logger.debug("Registered service is not capable of defining an mfa attribute.");
+            return null;
+        }
+
+        final ServiceMfaData serviceData = new ServiceMfaData();
+
+        final RegisteredServiceWithAttributes service = RegisteredServiceWithAttributes.class.cast(registeredService);
+
+        final Map mfaRole = Map.class.cast(service.getExtraAttributes().get("mfa_role"));
+        if (mfaRole == null) {
+            return null;
+        }
+
+        serviceData.setAttributeName(String.class.cast(mfaRole.get("mfa_attribute_name")));
+        serviceData.setAttributePattern(String.class.cast(mfaRole.get("mfa_attribute_pattern")));
+        serviceData.setAuthenticationMethod(String.class.cast(service.getExtraAttributes().get("authn_method")));
```

Comment:
Do we allow someone to specify a different key name for the services registry? That doesn't seem right, or at least it doesn't seem consistent with the JSON Services Registry... We don't allow them to define the name of the regex/ant expression used to match a service, or the allow to proxy attribute, does it make sense to do that here?

 (I can't tell you want it is called because it might change.) :)